src/utils/MetalBinding.mm

```diff
@@ -30,49 +33,154 @@
+                nxtSwapChainImplementation impl = {};
+                impl.Init = Init;
+                impl.Destroy = Destroy;
+                impl.Configure = Configure;
+                impl.GetNextTexture = GetNextTexture;
+                impl.Present = Present;
+                impl.userData = new SwapChainImplMTL(nswindow);
+                return impl;
+            }
+
+        private:
+            id nsWindow = nil;
+            id<MTLDevice> mtlDevice = nil;
+            id<MTLCommandQueue> commandQueue = nil;
+
+            CAMetalLayer* layer = nullptr;
+            id<CAMetalDrawable> currentDrawable = nil;
+            id<MTLTexture> currentTexture = nil;
+
+            SwapChainImplMTL(id nsWindow)
+                : nsWindow(nsWindow) {
             }
-            void GetProcAndDevice(nxtProcTable* procs, nxtDevice* device) override {
-                metalDevice = MTLCreateSystemDefaultDevice();
 
-                id nsWindow = glfwGetCocoaWindow(window);
+            ~SwapChainImplMTL() {
+                [currentTexture release];
+                [currentDrawable release];
+            }
+
+            void Init(nxtWSIContextMetal* ctx) {
+                mtlDevice = ctx->device;
+                commandQueue = [mtlDevice newCommandQueue];
+            }
+
+            nxtSwapChainError Configure(nxtTextureFormat format,
+                    uint32_t width, uint32_t height) {
+                if (format != NXT_TEXTURE_FORMAT_R8_G8_B8_A8_UNORM) {
+                    return "unsupported format";
+                }
+                ASSERT(width > 0);
+                ASSERT(height > 0);
+
                 NSView* contentView = [nsWindow contentView];
                 [contentView setWantsLayer: YES];
 
+                CGSize size = {};
+                size.width = width;
+                size.height = height;
+
                 layer = [CAMetalLayer layer];
-                [layer setDevice: metalDevice];
+                [layer setDevice: mtlDevice];
                 [layer setPixelFormat: MTLPixelFormatBGRA8Unorm];
                 [layer setFramebufferOnly: YES];
-                [layer setDrawableSize: [contentView bounds].size];
+                [layer setDrawableSize: size];
 
                 [contentView setLayer: layer];
 
-                backend::metal::Init(metalDevice, procs, device);
-                backendDevice = *device;
+                return NXT_SWAP_CHAIN_NO_ERROR;
+            }
 
-                backend::metal::SetNextDrawable(backendDevice, GetNextDrawable());
+            nxtSwapChainError GetNextTexture(nxtSwapChainNextTexture* nextTexture) {
+                [currentDrawable release];
+                currentDrawable = [layer nextDrawable];
+                [currentDrawable retain];
+
+                [currentTexture release];
+                currentTexture = currentDrawable.texture;
+                [currentTexture retain];
+
+                // Clear initial contents of the texture
```

Comment:
Instead of this, can we make BeginSubpass in CommandBufferMTL default to MTLLoadActionClear for color attachments? It might break the RenderToTexture sample but that's fine because we'll add load ops next.